Importer.pm

```diff
@@ -348,29 +339,92 @@ sub _prepareTrack {
 		REPLAYGAIN_ALBUM_PEAK => $album->{replay_peak},
 	};
 
+	$album->{release_type} = 'EP' if lc($album->{release_type} || '') eq 'epmini';
+
 	if ($album->{media_count} > 1) {
 		$attributes->{DISC} = $track->{media_number};
 		$attributes->{DISCC} = $album->{media_count};
 	}
 
-	if ($track->{composer} && $track->{composer}->{name}) {
+	if ( $track->{composer} && $track->{composer}->{name} && $track->{composer}->{name} !~ /^\s*various\s*composers\s*$/i ) {
 		$attributes->{COMPOSER} = $track->{composer}->{name};
+		$attributes->{COMPOSER_EXTID} = 'qobuz:artist:' . $track->{composer}->{id};
 		if ( $track->{work} && $prefs->get('importWorks') ) {
 			$attributes->{WORK} = $track->{work};
 		}
 	}
 
-	if ($track->{performer} && $artist !~ m/(^|\Q$_splitList\E)$track->{performer}->{name}($|\Q$_splitList\E)/i
-			&& Plugins::Qobuz::API::Common->trackPerformerIsMainArtist($track)) {
-		$attributes->{TRACKARTIST} = $track->{performer}->{name};
-	}
-
 	if ($track->{audio_info}) {
 		$attributes->{REPLAYGAIN_TRACK_GAIN} = $track->{audio_info}->{replaygain_track_gain};
 		$attributes->{REPLAYGAIN_TRACK_PEAK} = $track->{audio_info}->{replaygain_track_peak};
 	}
 
+	my ($artists, $artistIds);
+
+	foreach ( Plugins::Qobuz::API::Common->getMainArtists($album) ) {
+		push @{$artists}, $_->{name};
+		push @{$artistIds}, 'qobuz:artist:' . $_->{id};
+	}
+	if ( !$artists && $track->{performer} && $track->{performer}->{name} ) {
+		push @{$artists}, $track->{performer}->{name};
+		push @{$artistIds}, 'qobuz:artist:' . $track->{performer}->{id};
+	}
+
+	Plugins::Qobuz::API::Common->removeArtistsIfNotOnTrack($track, $artists, $artistIds);
+
+	if ($track->{performer} && Plugins::Qobuz::API::Common->trackPerformerIsMainArtist($track) && !grep $_ eq $track->{performer}->{name}, @{$artists}) {
+		push @{$artists}, $track->{performer}->{name};
+		push @{$artistIds}, 'qobuz:artist:' . $track->{performer}->{id};
+	}
+
+	$attributes->{ARTIST} = \@{$artists};
+	$attributes->{ARTIST_EXTID} = \@{$artistIds};
+	$allArtists{join(',', sort(@{$artistIds}))} = { names => \@{$artists}, ids => \@{$artistIds} };
+
+	my $rolePerformer;
+	if ($track->{performers}) {
+		my %seen;
+		my @performersAndRoles = split(' - ', $track->{performers});
+		foreach (@performersAndRoles) {
+			my %roleSeen = undef;
+			my @roles = split(/\s*,\s*/, $_);
+			my $name = shift @roles;
+			foreach (@roles) {
+				$_ =~ s/[\r$|\s*]//g;
+				$_ = uc($_);
+				push @{$rolePerformer->{$_}}, $name if !$roleSeen{$_};
+				$roleSeen{$_} = 1;
+			}
+		}
+		$attributes->{BAND} = $rolePerformer->{ORCHESTRA} if $rolePerformer->{ORCHESTRA};
+		$attributes->{CONDUCTOR} = $rolePerformer->{CONDUCTOR} if $rolePerformer->{CONDUCTOR};
```

Comment:
Use named variables inside nested loops, rather than the anonymous `$_`. This is prone to errors, as it changes, depending on the context. Do `foreach my $performarAndRoles (@performersAndRoles)` etc.